internal/handler/msg_hello.go

```diff
@@ -36,21 +40,115 @@ func (h *Handler) MsgHello(ctx context.Context, msg *wire.OpMsg) (*wire.OpMsg, e
 		return nil, lazyerrors.Error(err)
 	}
 
+	saslSupportedMechs, err := common.GetOptionalParam(doc, "saslSupportedMechs", "")
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+
+	var saslSupportedMechsResp *types.Array
+
+	if saslSupportedMechs != "" {
+		db, username, ok := strings.Cut(saslSupportedMechs, ".")
+		if !ok {
+			return nil, handlererrors.NewCommandErrorMsg(
+				handlererrors.ErrBadValue,
+				"UserName must contain a '.' separated database.user pair",
+			)
+		}
+		// If username is empty, MongoDB doesn't send back a saslSupportedMechs property but
+		// still sends the response normally.
+		if username != "" {
+			mechs, err := h.getUserSupportedMechs(ctx, db, username)
+			if err != nil {
+				return nil, lazyerrors.Error(err)
+			}
+
+			if len(mechs) > 0 {
+				saslSupportedMechsResp = must.NotFail(types.NewArray())
+				for _, k := range mechs {
+					saslSupportedMechsResp.Append(k)
+				}
+			}
+		}
+	}
+
+	resp := must.NotFail(types.NewDocument(
+		"isWritablePrimary", true,
+		"maxBsonObjectSize", int32(types.MaxDocumentLen),
+		"maxMessageSizeBytes", int32(wire.MaxMsgLen),
+		"maxWriteBatchSize", int32(100000),
+		"localTime", time.Now(),
+		"connectionId", int32(42),
+		"minWireVersion", common.MinWireVersion,
+		"maxWireVersion", common.MaxWireVersion,
+		"readOnly", false,
+	))
+
+	if saslSupportedMechsResp != nil {
+		resp.Set("saslSupportedMechs", saslSupportedMechsResp)
+	}
+
+	resp.Set("ok", float64(1))
+
 	var reply wire.OpMsg
-	must.NoError(reply.SetSections(wire.MakeOpMsgSection(
-		must.NotFail(types.NewDocument(
-			"isWritablePrimary", true,
-			"maxBsonObjectSize", int32(types.MaxDocumentLen),
-			"maxMessageSizeBytes", int32(wire.MaxMsgLen),
-			"maxWriteBatchSize", int32(100000),
-			"localTime", time.Now(),
-			"connectionId", int32(42),
-			"minWireVersion", common.MinWireVersion,
-			"maxWireVersion", common.MaxWireVersion,
-			"readOnly", false,
-			"ok", float64(1),
-		)),
-	)))
+	must.NoError(reply.SetSections(wire.MakeOpMsgSection(resp)))
 
 	return &reply, nil
 }
+
+// getUserSupportedMechs for a given user.
+func (h *Handler) getUserSupportedMechs(ctx context.Context, db, username string) ([]string, error) {
+	adminDB, err := h.b.Database("admin")
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+
+	usersCol, err := adminDB.Collection("system.users")
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+
+	filter, err := usersInfoFilter(false, false, db, []usersInfoPair{
+		{username: username, db: db},
+	})
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+
+	// Filter isn't being passed to the query as we are filtering after retrieving all data
+	// from the database due to limitations of the internal/backends filters.
```

Comment:
The same comment could be added to all `Query` calls, not only to this one; therefore, it adds nothing; that's just how `Query` works, and that's already documented.